clientapi/routing/register_test.go

```diff
@@ -264,3 +279,318 @@ func TestSessionCleanUp(t *testing.T) {
 		}
 	})
 }
+
+func Test_validateUsername(t *testing.T) {
+	tooLongUsername := strings.Repeat("a", maxUsernameLength)
+	tests := []struct {
+		name      string
+		localpart string
+		domain    gomatrixserverlib.ServerName
+		want      *util.JSONResponse
+	}{
+		{
+			name:      "empty username",
+			localpart: "",
+			domain:    "localhost",
+			want: &util.JSONResponse{
+				Code: http.StatusBadRequest,
+				JSON: jsonerror.InvalidUsername("Username can only contain characters a-z, 0-9, or '_-./='"),
+			},
+		},
+		{
+			name:      "invalid username",
```

Comment:
Say why (caps).

Also, it would be good to add tests for:
 - unicode characters e.g emoji
 - ASCII symbols e.g $
 - a complex but valid username beyond the existing test for "valid" e.g `f00_bar-baz.=40/` is valid I think? Test it.